### PR TITLE
Turn the `Visibility` struct into an enum

### PR DESCRIPTION
**Pull request by ickk**
# Objective

The current implementation of the `Visibility` struct simply wraps a boolean.. which seems like an odd pattern when rust has such nice enums that allow for more expression using pattern-matching.

## Solution

Make `Visibility` into an enum.

---

### Changed

`Visibility` is now an enum

## Migration Guide

- Code that checks the `visibility.is_visible` field should now call it as a method `visibility.is_inherited()`, use `==`, or refactor to use a `match`.
- Code that used to set or unset the `visibility.is_visible` field should now directly set the value: `*visibility = Visibility::Inherited`, or use the `.toggle()` or `.set(bool)` methods.
- Any usages of the associated constants `Visibility::VISIBLE` or `Visibility::INVISIBLE` should now use the enum variants, `Visibility::Inherited` or `Visibility::Hidden`.
- `ComputedVisibility::INVISIBLE` and `SpatialBundle::VISIBLE_IDENTITY` have been renamed to `ComputedVisibility::HIDDEN` and `SpatialBundle::INHERITED_IDENTITY` respectively.

### CODE REVIEW

**Comment by ickk** [comment]
Since the variants are self descriptive, we could `pub use` them like std does with `Option` so that users can avoid the `Visibility::` prefix.

**Comment by nicopap** [comment]
If we are open to a major breaking change, may I suggest a change that improves one of the issues I have with `Visibility` which is the stuttering. ie: `*visibility = Visibility::Invisible` (repeats "visible" three times)

What about:

```rust
enum Visibility {
  Show,
  Hide
}
```

This is already an improvement over `visibility.is_visible = Visibility { is_visible: false }` lol, but we still can do better.

Also I think here a `toggle`method is much needed for an enum-based API, since otherwise it would make it super annoying to swap the value (a bad good idea would be to implement `std::ops::Not`) (maybe two methods: one that returns a new value and one that mutably updates)

**Comment by ickk** [comment]
> Also I think here a `toggle` method is much needed for an enum-based API

see #6268 :)

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 1]
```diff
@@ -28,25 +28,35 @@ use crate::{
...
-    /// A [`Visibility`], set as visible.
-    pub const VISIBLE: Self = Visibility { is_visible: true };
+    /// Whether this entity is visible.
+    #[inline]
+    pub const fn is_visible(&self) -> bool {
+        match self {
+            Self::Visible => true,
+            Self::Invisible => false,
+        }
+    }
```
Comment (tguichaoua) [inline crates/bevy_render/src/view/visibility/mod.rs:50]: I suggest to use `Eq` instead of this kind of utility method for field-less enum.

**Review by tguichaoua — CHANGES_REQUESTED** [review]

**Comment by mockersf** [comment]
I'm not sure I think the `is_visible` method is really useful here

**Comment by sullyj3** [comment]
I'm happy to close #6268 in favour of this if there's a consensus that an Enum is a nicer api

**Comment by sullyj3** [comment]
I agree that `is_visible` feels redundant:
 
```rust
if vis == Visibility::Show { // ... }
if vis.is_visible() { //... }
```

The method is slightly terser, which is nice. Unless the variant was exported directly from the prelude:

```rust
if vis == Show { //... }
```
It would also depend on the variant names, eg in
```rust
if vis == Visibility::Visible { // ... }
if vis.is_visible() { //... }
```
the method seems much nicer.

`Option` has `.is_none()`, and that seems to be the idiomatic way to do things, despite the fact that you could just check `== None` there also.
Hmm, I guess to use `==` with `Option<T>`, `T` must be `PartialEq`, so `.is_some()` improves ergonomics in the case where it's not? Could that have been the rationale? We don't have that problem here.

I think I'm in favour of the method if the variants aren't exported unqualified, just because it's fewer characters.

**Comment by ickk** [comment]
I am more than happy to add an implementation for `Eq`, but I think the `is_visible` method should stay to provide a more consistent API with `ComputedVisibility`.

File: crates/bevy_render/src/view/visibility/mod.rs (hunk 1 above)
Comment (alice-i-cecile) [reply crates/bevy_render/src/view/visibility/mod.rs:50]: Agreed.

**Review by alice-i-cecile — COMMENTED** [review]
I'm in favor of using an enum here. Bikeshedding, I have a pretty strong preference for `Visibility::Show` and `Visibility::Hide`. Agree on adding `toggle`. `is_visible` can stay for consistency.

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 2]
```diff
@@ -26,31 +26,36 @@ use crate::{
...
-
-    /// A [`Visibility`], set as invisible.
-    pub const INVISIBLE: Self = Visibility { is_visible: false };
+    /// Whether this entity is visible.
+    #[inline]
+    pub const fn is_visible(&self) -> bool {
+        match self {
+            Self::Shown => true,
+            Self::Hidden => false,
+        }
```
Comment (mockersf) [inline crates/bevy_render/src/view/visibility/mod.rs:49]:
```suggestion
        matches!(self, Self::Shown)
```

**Comment by mockersf** [comment]
Could you impl the https://doc.rust-lang.org/std/ops/trait.Not.html trait and use it?

File: crates/bevy_render/src/view/visibility/mod.rs (hunk 2 above)
Comment (ickk) [reply crates/bevy_render/src/view/visibility/mod.rs]: 👍

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 3]
```diff
@@ -26,31 +27,42 @@ use crate::{
...
+    #[inline]
+    pub const fn is_visible(&self) -> bool {
+        matches!(self, Self::Shown)
+    }
 
     /// Toggle the visibility.
+    #[inline]
     pub fn toggle(&mut self) {
-        self.is_visible = !self.is_visible;
+        *self = !&*self
```
Comment (ickk) [inline crates/bevy_render/src/view/visibility/mod.rs:65]: @mockersf this `*self = !&*self` is the best I can do I think without adding `Copy`

Comment (alice-i-cecile) [reply crates/bevy_render/src/view/visibility/mod.rs:65]: I would totally add `Copy` to this struct :)

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 4]
```diff
@@ -26,31 +27,42 @@ use crate::{
...
 /// This is done by setting the values of their [`ComputedVisibility`] component.
-#[derive(Component, Clone, Reflect, Debug)]
+#[derive(Component, Clone, Copy, Reflect, Debug, PartialEq, Eq)]
 #[reflect(Component, Default)]
-pub struct Visibility {
-    /// Indicates whether this entity is visible. Hidden values will propagate down the entity hierarchy.
-    /// If this entity is hidden, all of its descendants will be hidden as well. See [`Children`] and [`Parent`] for
-    /// hierarchy info.
-    pub is_visible: bool,
+pub enum Visibility {
```
Comment (james7132) [inline crates/bevy_render/src/view/visibility/mod.rs:32]: Ensure the size of the type does not change via `repr(u8)`.

**Review by james7132 — CHANGES_REQUESTED** [review]
Not sure I agree with this decision. While this improves simple assignments and serialized formats, this does make it notably harder to assign visibility state with even simple boolean expressions, instead requiring match/if expressions, which can be a lot less ergonomic.

**Review by mockersf — APPROVED** [review]

**Comment by sullyj3** [comment]
> this does make it notably harder to assign visibility state with even simple boolean expressions, instead requiring match/if expressions, which can be a lot less ergonomic.

What if there was a `Visibility::from_bool()`?

File: crates/bevy_render/src/view/visibility/mod.rs (hunk 4 above)
Comment (ickk) [reply crates/bevy_render/src/view/visibility/mod.rs:32]: From what I understand, the compiler should already use a `u8` for this enum, and in fact asking for it to unconditionally use a `u8` means the compiler won't be able to make [optimisations](https://github.com/rust-lang/rust/pull/45225) when you start nesting the type inside of other values like in an `Option` or `Result` (I could be wrong though, maybe the compiler is free to ignore the `repr(u8)` under certain circumstances?).

**Comment by ickk** [comment]
I agree could be more terse when modifying an existing instance using the result of a boolean expression.

I wrote out all the scenarios so I could evaluate using From<bool>/Into, and I did not really love it.

We could discuss adding something like a `set_when` method or something maybe?
```rust
visibility.set_when(a && b || !c);
```

**Comment by alice-i-cecile** [comment]
Why not just `visibility.set(visibility.set_when(a && b || !c));` where the `set` method takes a `bool`?

File: crates/bevy_render/src/view/visibility/mod.rs (hunk 4 above)
Comment (james7132) [reply crates/bevy_render/src/view/visibility/mod.rs:32]: If that is the case, please add a test to ensure that the size remains at 1 byte for both `Visibility` and `Option<Visibility>` . The performance of many rendering systems is reliant on this type staying small.

File: crates/bevy_render/src/view/visibility/mod.rs [hunk 5]
```diff
@@ -26,31 +27,52 @@ use crate::{
...
+    Inherited,
```
Comment (james7132) [inline crates/bevy_render/src/view/visibility/mod.rs:33]: Please add per variant docs on what behavior each variant specifies. Particularly with Inherited and how it's operates on root nodes.

**Review by james7132 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**